controllers/hbase_controller.go

```diff
@@ -587,24 +591,28 @@ func (r *HBaseReconciler) ensureStatefulSet(hb *hbasev1.HBase,
 		}
 		return nil, false, err
 	}
-	r.Log.Info("updating revision", "sameRevision", expectedRevision != actual.Annotations[HBaseControllerRevisionKey])
-	r.Log.Info("updating replica count", "sameReplicaCount", *actual.Spec.Replicas != *expected.Spec.Replicas)
 
-	if actual.Annotations != nil &&
-		actual.Annotations[HBaseControllerRevisionKey] == expectedRevision &&
-		*actual.Spec.Replicas == *expected.Spec.Replicas {
-		r.Log.Info("StatefulSet is up-to-date", "name", stsName)
-		return actual, false, nil
+	var actualRevision string
+	if actual.Annotations != nil {
+		actualRevision = actual.Annotations[HBaseControllerRevisionKey]
 	}
 
-	// update
-	if err := r.Update(context.TODO(), expected); err != nil {
-		r.Log.Info("StatefulSet is up-to-date", "name", stsName)
-		return nil, false, err
+	r.Log.Info("StatefulSet reconciliation",
+		"name", stsName,
+		"revision is up to date", actualRevision == expectedRevision,
+		"expected replica count", *actual.Spec.Replicas == *expected.Spec.Replicas)
+
+	if *actual.Spec.Replicas != *expected.Spec.Replicas || actualRevision != expectedRevision {
+		// Update StatefulSet to expected configuration
+		if err := r.Update(context.TODO(), expected); err != nil {
```

Comment:
What it happening from my point of view **before** this change:
- ensureStatefulSet() updates StatefulSet
  - get expected state with statefulSet()
    - The function makes StatefulSet object, compute hash of the object and set a hash attribute for it
    - Hash is computed without replicas set, so changing number of replicas did not change hash
    - Hash is also computed without `last-applied-configuration` annotation too, so hash do not depend on it
    - Function specifically set `last-applied-configuration` annotation to the pods template of the StatefulSet. This annotation copied from hbase CRD.
    - Because of that change only # of replicas do not change StatefulSet hash but change annotation of pods template.
  - Stateful set is updated. In the case of any changes in the StatefulSet reconciler return and retry again from the beginning.
- Asynchronously StatefulSet controller noticed changes of the StatefulSet resource
  - For pod updates tracking it uses its own hash `controller-revision-hash` (completely different from hbase hashing)
  - AFAIK it computes `controller-revision-hash` from pod template spec from the StatefulSet resource
  - Because statefulSet() added `last-applied-configuration` annotation, pod template hash changed and controller set new hash as `update revision`.
  - Controller can start new pods or delete pods in the case of # of replicas change, but `OnDelete` policy prevents controller from update pods revision.
- Asynchronously Hbase controller restart reconciliation and run ensureStatefulSetPods()
  - It compares `update revision` from StatefulSet and `controller-revision-hash` for every individual pods.
  - It starts rolling restart for pods when `update revision` != `controller-revision-hash`

There are two problems:
- Pod template annotation changed when only # of replicas changed. It is fixed with the PR (using filtered annotation)
- There is a possible race condition between StatefulSet controller and Hbase controller. It can cause different issues including the situation when rolling restart does not happen even without the fix.

Reply:
Thanks for the in-depth debugging - makes sense. Approving change --